codegen/module_system.go

```diff
@@ -0,0 +1,199 @@
+		ClassType:         module.MultiModule,
+		ClassDependencies: []string{},
+	}); err != nil {
+		return nil, errors.Wrapf(err, "Error registering client class")
+	}
+
+	if err = system.RegisterClassType("client", "http", &HTTPClientGenerator{
+		templates:     tmpl,
+		packageHelper: h,
+	}); err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"Error registering http client class type",
+		)
+	}
+
+	// TODO: Register endpoint module class and type generators
+	return system, nil
+}
+
+/*
+ * HTTP Client Generator
+ */
+
+// HTTPClientGenerator generates an instance of a zanzibar http client
+type HTTPClientGenerator struct {
+	templates     *Template
+	packageHelper *PackageHelper
+}
+
+// Generate returns the HTTP client generated files as a map of relative file
+// path (relative to the target buid directory) to file bytes.
+func (generator *HTTPClientGenerator) Generate(
+	instance *module.Instance,
+) (map[string][]byte, error) {
+	// Parse the client config from the endpoint JSON file
+	clientConfig, err := generator.readConfig(instance.JSONFileRaw)
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"Error reading HTTP client %s JSON config",
+			instance.InstanceName,
+		)
+	}
+
+	clientSpec, err := NewHTTPClientSpec(
+		filepath.Join(
+			instance.BaseDirectory,
+			instance.Directory,
+			instance.JSONFileName,
+		),
+		clientConfig,
+		generator.packageHelper,
+	)
+	if err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"Error initializing HTTPClientSpec for %s",
+			instance.InstanceName,
+		)
+	}
+
+	clientMeta := &ClientMeta{
+		PackageName:      clientSpec.ModuleSpec.PackageName,
+		Services:         clientSpec.ModuleSpec.Services,
+		IncludedPackages: clientSpec.ModuleSpec.IncludedPackages,
+		ClientID:         clientSpec.ClientID,
+	}
+
+	// TODO: add helper to template to return byte array
+	clientBuffer := bytes.NewBuffer(nil)
+	clientWriter := bufio.NewWriter(clientBuffer)
+	if err := generator.templates.template.ExecuteTemplate(
+		clientWriter,
+		"http_client.tmpl",
+		clientMeta,
+	); err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"Error executing HTTP Client template for %s",
+			instance.InstanceName,
+		)
+	}
+	if err := clientWriter.Flush(); err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"Error generating HTTP client from template for %s",
+			instance.InstanceName,
+		)
+	}
+
+	structsBuffer := bytes.NewBuffer(nil)
+	structsWriter := bufio.NewWriter(structsBuffer)
+	if err := generator.templates.template.ExecuteTemplate(
+		structsWriter,
+		"structs.tmpl",
+		clientMeta,
+	); err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"Error executing HTTP client structs template for %s",
+			instance.InstanceName,
+		)
+	}
+	if err := structsWriter.Flush(); err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"Error generating HTTP client structs from template for %s",
+			instance.InstanceName,
+		)
+	}
+
+	clientDirectory := filepath.Join(
+		instance.BaseDirectory,
+		"build",
```

Comment:
This is hard coded here but actually comes from gateway.json.

We should avoid duplicate source of truth, either zanzibar respects "targetGenDir" ( https://github.com/uber/zanzibar/blob/master/examples/example-gateway/gateway.json#L6 ) or is always hard coded to write to "./build/"